rust/connlib/snownet/src/node.rs

```diff
@@ -1400,31 +1389,109 @@ enum ConnectionState<RId> {
+    Idle {
+        /// Our nominated socket.
+        peer_socket: PeerSocket<RId>,
     },
     /// The connection failed in an unrecoverable way and will be GC'd.
     Failed,
-    /// The connection is idle and will be GC'd.
-    Idle,
 }
 
-impl<RId> ConnectionState<RId> {
-    fn add_possible_socket(&mut self, socket: SocketAddr) {
-        let possible_sockets = match self {
-            ConnectionState::Connecting {
-                possible_sockets, ..
-            } => possible_sockets,
+impl<RId> ConnectionState<RId>
+where
+    RId: Copy,
+{
+    fn poll_timeout(&self) -> Option<Instant> {
+        match self {
             ConnectionState::Connected {
-                possible_sockets, ..
-            } => possible_sockets,
-            ConnectionState::Idle | ConnectionState::Failed => return,
+                last_incoming,
+                last_outgoing,
+                ..
+            } => Some(idle_at(*last_incoming, *last_outgoing)),
+            ConnectionState::Connecting { .. }
+            | ConnectionState::Idle { .. }
+            | ConnectionState::Failed => None,
+        }
+    }
+
+    fn handle_timeout(&mut self, agent: &mut IceAgent, now: Instant) {
+        let Self::Connected {
+            last_outgoing,
+            last_incoming,
+            peer_socket,
+        } = self
+        else {
+            return;
         };
 
-        possible_sockets.insert(socket);
+        if idle_at(*last_incoming, *last_outgoing) > now {
+            return;
+        }
+
+        let peer_socket = *peer_socket;
+
+        self.transition_to_idle(peer_socket, agent);
+    }
+
+    fn on_outgoing(&mut self, agent: &mut IceAgent, now: Instant) {
+        let peer_socket = match self {
+            Self::Idle { peer_socket } => *peer_socket,
+            Self::Connected { last_outgoing, .. } => {
+                *last_outgoing = now;
+                return;
+            }
+            Self::Failed | Self::Connecting { .. } => return,
+        };
+
+        self.transition_to_connected(peer_socket, agent, now);
+    }
+
+    fn on_incoming(&mut self, agent: &mut IceAgent, now: Instant) {
+        let peer_socket = match self {
+            Self::Idle { peer_socket } => *peer_socket,
+            Self::Connected { last_incoming, .. } => {
+                *last_incoming = now;
+                return;
+            }
+            Self::Failed | Self::Connecting { .. } => return,
+        };
+
+        self.transition_to_connected(peer_socket, agent, now);
+    }
+
+    fn transition_to_idle(&mut self, peer_socket: PeerSocket<RId>, agent: &mut IceAgent) {
+        tracing::debug!("Connection is idle");
+        *self = Self::Idle { peer_socket };
+        apply_idle_stun_timings(agent);
+    }
+
+    fn transition_to_connected(
+        &mut self,
+        peer_socket: PeerSocket<RId>,
+        agent: &mut IceAgent,
+        now: Instant,
+    ) {
+        tracing::debug!("Connection is no longer idle");
```

Comment:
Open to better wordings here, I couldn't think of a wording that is a nice "opposite" of "idle".